alt-tab-macos/ui/PreferencesPanel.swift

```diff
@@ -1,110 +1,211 @@
+
+        return makeLabelWithProvidedControl(labelText, rawName, popUp, suffixText)
+    }
+
+    private func makeLabelWithSlider(_ keyPath: ReferenceWritableKeyPath<PreferencesPanel, NSSlider?>, _ labelText: String, rawName: String, minValue: Double, maxValue: Double, numberOfTickMarks: Int, suffixText: String? = nil) -> NSStackView {
+        let slider = NSSlider(
+                value: Double(Preferences.rawValues[rawName]!)!,
+                minValue: minValue,
+                maxValue: maxValue,
+                target: self,
+                action: #selector(controlDidEndEditing)
+        )
+        slider.numberOfTickMarks = numberOfTickMarks
+        slider.allowsTickMarkValuesOnly = numberOfTickMarks > 1
+        slider.tickMarkPosition = .below
+
+        self[keyPath: keyPath] = slider
+
+        return makeLabelWithProvidedControl(labelText, rawName, slider, suffixText)
+    }
+
+    private func makeLabelWithProvidedControl(_ labelText: String, _ rawName: String, _ control: NSControl, _ suffixText: String? = nil) -> NSStackView {
+        let label = NSTextField(wrappingLabelWithString: labelText + ": ")
+        label.alignment = .right
+        label.widthAnchor.constraint(equalToConstant: calcLabelWidth()).isActive = true
+
+        control.identifier = NSUserInterfaceItemIdentifier(rawName)
+        control.target = self
+        control.action = #selector(controlDidEndEditing)
+        let containerView = NSStackView(views: [label, control])
+
+        if suffixText != nil {
+            let suffix = NSTextField(labelWithString: suffixText!)
+            suffix.textColor = .gray
+            containerView.addView(suffix, in: .leading)
         }
+
+        return containerView
     }
 
-    func textDidEndEditing(_ notification: Notification) {
-        if let textView = notification.object as? NSTextView {
-            let key = inputsMap[textView]!
+    /*
+    usage notes:
+    - NSSlider: supports on purpose currently only decimal values
+    */
+    @objc func controlDidEndEditing(senderControl: NSControl) {
+        self.makeFirstResponder(invisibleTextField) // deselects any possibly selected NSTextField (so slider & popUp changes deselect them)
+
+        let key: String? = senderControl.identifier?.rawValue
+        var newValue: String?
+
+        if senderControl is NSPopUpButton {
+            newValue = (senderControl as! NSPopUpButton).titleOfSelectedItem!
+        } else if senderControl is NSSlider {
+            newValue = String(format: "%.0f", Double(senderControl.stringValue)!) // we are only interested in decimals of the provided double
+        } else {
+            newValue = senderControl.stringValue
+        }
+
+//        debugPrint("PreferencesPanel: save: change", key!, newValue!)
+
+        if key != nil && newValue != nil {
+            if newValue == Preferences.rawValues[key!] {
+                debugPrint("PreferencesPanel: save: abort: value was not changed")
```

Comment:
I think there is a bug here. I keep seeing "PreferencesPanel: save: abort: value was not changed" in the logs when I use the max width slider and move to any position

Reply:
The slider reported back way to often to `controlDidEndEditing`: mouseDown, mouseMove, mouseUp. It got the latest user selected value already with mouseMove (and saved that one) and skipped the value from the mouseUp event.

This was one of the reasons for me adding the `newValue` & `previousValue` comparison. The other was to prevent unneeded saves.

I found an interesting attribute in `NSSlider` which I disabled now: `isContinuous`. This way it only calls the `controlDidEndEditing` callback on mouseUp 💯